### ISSUE TITLE
oob_prediction_ in RandomForestClassifier

### ISSUE THREAD

**Issue by WZheng-94** [UC-MACSS/persp-model_W18#267]
Does any one know how to get error rates for each category of a binary variable in RandomForestClassifier? 
I found out obb_prediction_  seemed to be exclusive to RandomForestRegressor. 

Thanks.

**Comment by khan1792** [comment]
Combine the y values and the oob prediction values into a data frame
Split it into two groups (prediction >=0.5 or not)
Respectively calculate the Bayes error rate for two groups

**Comment by ZhengErWei** [comment]
But the problem set asks us to use RandomForestClassifier?

**Comment by khan1792** [comment]
I mistakenly used the RFregression. But the calculation method remain the same. I just took one more step to transform the prediction numeric values as 1 or 0, while RFclassifier would directly produce them.

**Comment by ZhengErWei** [comment]
My problem is I think RFclassifier doesn't have oob_prediction_  attribute. I hope it's somewhere I missed.

**Comment by khan1792** [comment]
Ahhhh, I see... haven’t tried it...

**Comment by khan1792** [comment]
I have check the original code of sklearn package. In RandomForestClassifier, we can use oob_decision_function_ to calculate the oob prediction. 
1. Transpose the matrix produced by oob_decision_function_
2. Select the second raw of the matrix
3. Set a cutoff and transform all decimal values as 1 or 0 (>= 0.5 is 1 and otherwise 0)
The list of values we finally get is the oob prediction.

**Comment by ZhengErWei** [comment]
Thank you so much! I will give it a shot later.

**Comment by AlexanderTyan** [comment]
Cross checking against the way sklearn calculates oob_score_, I believe cutoff for step 3 should be:
 (>0.5 is 1 and <=0 is 0). This may impact MSE score calculations.

**Comment by khan1792** [comment]
It seems that the oob_score_ function in sklearn has a serious problem both in RF regression and classification. Especially in the former one, the oob score function sometimes produces weird results. The principle of out of bag estimation is very clear. We estimate a decision tree model via the bootstrapped training sample (averagely 2/3 of the original sample), and predict the oob values via bootstrapped test sample. We finally average oob values in regression or classify them in classification. I've never seen any statistical book sets a cut off that ignores the range (0 - 0.5) or there is a predicted value in classification that is between 0 - 0.5. You can check [the original code of oob_score here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/forest.py), it is a bit weird.

**Comment by khan1792** [comment]
Forget the last message... 
Sure, I didn't realize there is a 0.5 in the oob prediction. The oob_score_ classifies 0.5 as 0 and I classified it as 1. This cutoff is tricky in small data, although the classification decision is arbitrary.

**Comment by AlexanderTyan** [comment]
Right, I think 0.5 IS quite arbitrary on the part of sklearn. I just went along with it for MSE calculations because if that's what sklearn decides to come up with predicted y values, it would consistent to evaluate MSE according to the same thresholds.

**Comment by AlexanderTyan** [comment]
Looking to why the 0.5, I found:
https://books.google.com/books?id=bRpYDgAAQBAJ&pg=PT278&lpg=PT278&dq=sklearn+oob_decision_function_&source=bl&ots=h3VyS5rM9K&sig=Yw0RkHTA4whxi8E5teWMiF6bX6o&hl=en&sa=X&ved=0ahUKEwj-rfu519TZAhWl24MKHf-7D8YQ6AEIjQEwCA#v=onepage&q=sklearn%20oob_decision_function_&f=false

It seems that the way sklearn decides if category 0 or 1 is it looks at the relative probabilities of an observation is 0 or 1 in the the  oob_decision_function_ matrix.  Of second value is larger than the first, it decides in favour of 1.  Since probability adds up to one, this effectively means the 0.5 threshold.

**Comment by khan1792** [comment]
Most classifiers decide the value according to this kind of probability. Theoritically the thread point should not be classified as 1 or 0 because it will make both sides unbalanced. But practically it’s not a problem no matter it is classified as 1 or 0. As long as the sample size is large enough and the number of tree we set in the model is also large, It’s effect on the result will decrease.
